test/ConnectedRouter.test.js

```diff
@@ -135,6 +135,126 @@ describe('ConnectedRouter', () => {
 
       expect(onLocationChangedSpy.mock.calls[1][0].state).toEqual({ foo: 'bar'})
     })
+    
+    it('changing store location updates history', () => {
+      store = createStore(
+        combineReducers({
+          router: connectRouter(props.history)
+        }),
+        compose(applyMiddleware(routerMiddleware(props.history)))
+      )
+      
+      mount(
+        <Provider store={store}>
+          <ConnectedRouter {...props}>
+            <Route path="/" render={() => <div>Home</div>} />
+          </ConnectedRouter>
+        </Provider>
+      )
+
+      // Need to add PUSH action to history because initial POP action prevents history updates
+      props.history.push({ pathname: "/" })
+
+      store.dispatch({
+        type: LOCATION_CHANGE,
+        payload: {
+          location: {
+            pathname: '/',
+            search: '',
+            hash: '',
+            state: { foo: 'bar' }
+          },
+          action: 'PUSH',
+        }
+			})
+			
+			expect(props.history.entries).toHaveLength(3)
+
+      store.dispatch({
+        type: LOCATION_CHANGE,
+        payload: {
+          location: {
+            pathname: '/',
+            search: '',
+            hash: '',
+            state: { foo: 'bar' }
+          },
+          action: 'PUSH',
+        }
+			})
+			
+			expect(props.history.entries).toHaveLength(3)
+			
+      store.dispatch({
+        type: LOCATION_CHANGE,
+        payload: {
+          location: {
+            pathname: '/',
+            search: '',
+            hash: '',
+            state: { foo: 'baz' }
+          },
+          action: 'PUSH',
+        }
+      })
+
+      expect(props.history.entries).toHaveLength(4)
+    })
+    
+    it('supports custom state compare function', () => {
+      store = createStore(
+        combineReducers({
+          router: connectRouter(props.history)
+        }),
+        compose(applyMiddleware(routerMiddleware(props.history)))
+      )
+      
+      mount(
+        <Provider store={store}>
+          <ConnectedRouter
+            stateCompareFunction={(a, b) => {
+							return a === undefined  || (a.foo === "baz" && b.foo === 'bar') ? true : false
```

Comment:
No need to add `a === undefined` which we doesn't test it. Only the second comparison is enough to test this comparison feature.

```suggestion
              return (a.foo === 'baz' && b.foo === 'bar') ? true : false
```